chain/chunks/src/lib.rs

```diff
@@ -1561,6 +1559,28 @@ impl ShardsManager {
         Ok(())
     }
 
+    /// Processes the forwarded chunk parts for the header, and if the chunk is complete,
+    /// processes the chunk.
+    pub fn process_cached_chunk_forwards_for_header(
+        &mut self,
+        header: &ShardChunkHeader,
+        chain_store: &mut ChainStore,
+        rs: &mut ReedSolomonWrapper,
+    ) -> Result<ProcessPartialEncodedChunkResult, Error> {
+        if self.encoded_chunks.get_or_insert_from_header(header).complete {
+            return Ok(ProcessPartialEncodedChunkResult::Known);
+        }
+        if let Some(parts) = self.chunk_forwards_cache.pop(&header.chunk_hash()) {
+            self.encoded_chunks.merge_in_partial_encoded_chunk(&PartialEncodedChunkV2 {
```

Comment:
You need to check the merkle root of the chunk forward is the same as the merkle root of the chunk header

Reply:
I think this is validated already (this function is called by the client before inserting the forward):
```
    pub fn validate_partial_encoded_chunk_forward(
        &mut self,
        forward: &PartialEncodedChunkForwardMsg,
    ) -> Result<(), Error> {
        let valid_hash = forward.is_valid_hash(); // check hash

        if !valid_hash {
            return Err(Error::InvalidPartMessage);
        }

        // check part merkle proofs
        let num_total_parts = self.runtime_adapter.num_total_parts();
        for part_info in forward.parts.iter() {
            self.validate_part(forward.merkle_root, part_info, num_total_parts)?;
        }
```
forward.is_valid_hash() does
```
    pub fn is_valid_hash(&self) -> bool {
        let correct_hash = combine_hash(&self.inner_header_hash, &self.merkle_root);
        ChunkHash(correct_hash) == self.chunk_hash
    }
```
which verifies that the merkle_root does hash to the chunk_hash, and since the chunk_hash is used to index the forward, this should be fine.

Reply:
I see, good observation. Can you add some comments here explaining why the forward messages don't need to be validated again?